proposals/0000-hypercore-header.md

```diff
@@ -0,0 +1,108 @@
+
+
+# Motivation
+[motivation]: #motivation
+
+At time of writing, Dat's standard data structures are being expanded upon (hyperdb) while existing data structures are being updated with breaking changes (hyperdrive). In order for these changes to be deployed smoothly to the network, it's important that the following needs be met:
+
+ 1. Programs should correctly identify the data structure type and version within a hypercore. This will enable the program to use the correct handling code (eg hyperdb, hyperdrive-v1, hyperdrive-v2).
+ 2. Programs should abort processing a hypercore if they lack support for the type or version of the data structure.
+
+
+# Usage Documentation
+[usage-documentation]: #usage-documentation
+
+The "header" is the first entry in a hypercore. It includes a `type` and an optional `extension` which can contain data-structure-specific data.
+
+A program that is reading a hypercore will examine the `type` to determine how to process the hypercore.
+
+```
+function loadCorrectStructure (hypercore) {
+  
+  var header = parseHypercoreHeader(hypercore.readEntry(0))
+
+  if (!header) {
+    // no header present - treat as a hypercore
+    return hypercore
+  }
+
+  switch (header.type) {
+    case 'hyperdrive':
+      return new HyperdriveV1(hypercore, header)
+    case 'hyperdrive-v2':
+      return new HyperdriveV2(hypercore, header)
+    case 'hyperdb':
+      return new HyperdbV1(hypercore, header)
+    // ...
+    default:
+      // unknown type - treat as a hypercore
+      return hypercore
+  }
+
+}
+```
+
+The `type` string should be unique to the data structure (refer to existing DEPs to avoid conflicts). When a breaking change is made to the data-structure, it should be given a new `type` string. For example, the `"hyperdrive"` type string will be updated to `"hyperdrive-v2"` for its second version.
+
+The `header.extension` is an optional blob of bytes. It can be used by the data structure to store custom data. In `"hyperdrive"` (v1) it is used to specify the key of the content hypercore. It is also valid to encode a protobuf message into the extension field, and therefore it's simple to add additional fields.
+
+
+# Reference Documentation
+[reference-documentation]: #reference-documentation
+
+The header message should be written as the first entry of the hypercore. It is not required that all hypercores possess a header.
+
+The header message has the following protobuf schema:
+
+```protobuf
+message HypercoreHeader {
+  required string type = 1;
```

Comment:
I'm starting to wonder if we should call this `dataStructure` instead of `type`, because "type" is a really generic word that could be used elsewhere in the stack while "data structure" is very specific to what this is.

Reply:
Noting that this can be updated without any backwards compat issues because of protobuf

Reply:
`type` is very generic. I'd bikeshed to `contentType` or `contentSchema`, but loosely held opinion (`dataStructure` works for me).